GSoC 2018 Approximate convex decomposition

liubsp opened the pull request:
The algorithm in this project was proposed by Khaled Mamou and Faouzi Ghorbel in the paper:
A simple and efficient approach for 3d mesh approximate convex decomposition. Groupe de recherche GRIFT, 2010.

It’s an efficient approach to decompose a concave 3D mesh into a set of nearly convex surfaces and is useful for skeleton extraction, patterns recognition applications and particularly for the increase of collision detection accuracy without significantly affecting its performance.

liubsp committed:
Created the project's folder

liubsp committed:
Add external API, the base of internal implementaion, and examples

sloriot commented on Approximate_convex_decomposition/include/CGAL/approx_decomposition.h:
```diff
@@ -0,0 +1,113 @@
+#ifndef CGAL_APPROXIMATE_CONVEX_DECOMPOSITION_APPROX_DECOMPOSITION_H
+#define CGAL_APPROXIMATE_CONVEX_DECOMPOSITION_APPROX_DECOMPOSITION_H
+
+#define CGAL_DEFAULT_CONCAVITY_THRESHOLD 0.5
```
we try to avoid using macros when they are not needed

liubsp replied:
Isn't it a bad practice to put the same constants everywhere? Maybe, something like constexpr double CGAL_DEFAULT_CONCAVITY_THRESHOLD = 0.5;  would be better?

sloriot replied:
this value should not be spread all around, it should actually be at one place where the named parameter of the function is recovered so it is not needed to have it globally defined.

liubsp replied:
![image](https://user-images.githubusercontent.com/12598936/40534520-f6d14f3a-600e-11e8-8c30-d4d5bdb52906.png)
Just look at those repeating numbers of number_of_clusters, smoothing_lambda, output_cluster_ids.
In case, when you want to change any of those, you have to change them at each function declaration.

sloriot replied:
The discussion is meaningless provided no serious code doing something exists yet :)

liubsp replied:
But IMO it's better to define some rules before writing any serious code :)

sloriot commented on Approximate_convex_decomposition/include/CGAL/approx_decomposition.h:
```diff
@@ -0,0 +1,113 @@
+           class GeomTraits = typename Kernel_traits<typename TriangleMesh::Point_3>::Kernel
+           >
+ void
+ convex_decomposition_clusters(const TriangleMesh& mesh,
```
what is a cluster mesh? If this is a part of a mesh, you might look at [segmentation example](https://doc.cgal.org/latest/Surface_mesh_segmentation/#title14) to see how a submesh can be extracted.

liubsp replied:
At each iteration of the algorithm, a subset of connected facets corresponds to a single cluster, all missing parts are triangulated to make it closed, and after that, the cluster is produced. In practice, lots of the facets will not belong to any cluster. You might watch the video that shows up how it works (especially, look at iterations where the number of clusters is small; you will see that cluster is not necessary a submesh):
https://www.youtube.com/watch?v=T0U0_9M9LTw

Is in CGAL any tool to automatically triangulate missing parts of a mesh?

sloriot replied:
IMO, it is simply the convex hull of the faces per cluster that is depicted in the video. You don't need to produce any meshes.

liubsp replied:
Okay, I understand your point and will update the API.

liubsp replied:
I have commited new changes.

liubsp committed:
Update to API: remove cluster meshes

sloriot commented on Approximate_convex_decomposition/include/CGAL/approx_decomposition.h:
```diff
@@ -0,0 +1,64 @@
+                     double concavity_threshold = 100,
+                     std::size_t min_number_of_clusters = 1,
+                     PointPropertyMap ppmap = PointPropertyMap(),
+                     const GeomTraits& traits = GeomTraits())
```
we'll probably use named parameters at the end but that does not matter for the moment, the critical part is the algorithm.

liubsp committed:
Add the implementations of concavity value and dual graph

liubsp committed:
Build fix (an uncommented problematic line)

liubsp committed:
Add first iteration implementation of the algorithm

liubsp committed:
Fix bugs: first working but slow version

sloriot commented:
Please update your master branch so that only yours commits are displayed. This will make it easier to review.

liubsp committed:
Clean the code and update the comments, add working concavity example

CUBERLEON commented:
Okay, but I have no idea how to remove that merge commit yet.

sloriot commented:
I think if you update you master branch from cgal one it should be good.
something like `git push mine cgal/master:master`

CUBERLEON commented:
All fixed.

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+    typedef typename GeomTraits::Point_3 Point_3;
+    typedef typename GeomTraits::Vector_3 Vector_3;
+ 
+    typedef CGAL::Surface_mesh<Point_3> Surface_mesh;
```
line to be removed?

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+ 
+    typedef CGAL::Surface_mesh<Point_3> Surface_mesh;
+
+    typedef CGAL::Face_filtered_graph<Surface_mesh> Face_filtered_mesh;
```
line to be removed?

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+    , m_traits(traits)
+    {}
+
+    ~Approx_decomposition()
```
useless

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+
+            if (!found_edge) break;
+            
+//            std::cout << std::endl;
```
If you want to keep those you can enclose them in `#ifdef CGAL_CONVEX_DECOMPOSITION_VERBOSE` for example

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+    template <class Mesh>
+    struct Noborder_predicate
+    {
+        Noborder_predicate() : m_mesh(NULL) {}
```
will not compile so you can simply remove it I guess

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/concavity.h:
```diff
@@ -0,0 +1,143 @@
+            return compute(conv_hull);
+        }
+
+        double compute(const Surface_mesh& conv_hull)
```
TriangleMesh?

sloriot commented on Approximate_convex_decomposition/include/CGAL/approx_decomposition.h:
```diff
@@ -0,0 +1,87 @@
+#include <CGAL/boost/graph/helpers.h>
+#include <CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h>
+#include <CGAL/internal/Approximate_convex_decomposition/concavity.h>
+#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
```
you should not include the kernel. If you want to have a default, you can use `Kernel_traits` to use the kernel of the points associated with the vertex point map.

liubsp replied:
But there is no vertex point map.

sloriot replied:
There should be :)
I suggest to look at a function from polygon mesh processing that is using named parameters to see how you can do it. For example you can write something like that:
```
template <class TriangleMesh , class NamedParameters>
void toto( const TriangleMesh& pmesh
              , const NamedParameters& np)
{
  using boost::choose_param;
  using boost::get_param;
  
  typename GetGeomTraits<PolygonMesh, NamedParameters>::type Traits;
  typename GetVertexPointMap<PolygonMesh, NamedParameters>::const_type
    vpm = choose_param(get_param(np, internal_np::vertex_point),
                       get_const_property_map(CGAL::vertex_point, pmesh));
}
```

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+#include <vector>
+#include <utility>
+
+#include "concavity.h"
```
please update include directive

sloriot commented on Approximate_convex_decomposition/examples/Approximate_convex_decomposition/approx_convex_decomposition_SM_example.cpp:
```diff
@@ -0,0 +1,69 @@
+    // create property map for cluster-ids
+    typedef std::map<face_descriptor, int> Facet_int_map;
+    Facet_int_map facet_map;
+    boost::associative_property_map<Facet_int_map> facet_property_map(facet_map);
```
the point of using Surface_mesh is that you can have some internal properties. See `add_property_map` in the doc.

sloriot commented on Approximate_convex_decomposition/examples/Approximate_convex_decomposition/approx_convex_decomposition_example.cpp:
```diff
@@ -0,0 +1,55 @@
+    // create property map for cluster-ids
+    typedef std::map<Polyhedron::Facet_const_handle, int> Facet_int_map;
+    Facet_int_map facet_map;
+    boost::associative_property_map<Facet_int_map> facet_property_map(facet_map);
```
for polyhedron you can use dynamic properties. See the [CGAL BGL package](https://doc.cgal.org/latest/BGL/index.html#title25).

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,363 @@
+        }
+
+        Surface_mesh new_cluster;
+        Face_filtered_mesh selected_faces(m_mesh, decimation_props.new_cluster_props.faces);
```
why do you need a Face_filtered_graph here? I'm surprized it compiles with the default Polyhedron type?

liubsp replied:
Each cluster contains a list of faces that corresponds to the cluster. I construct new mesh using that list and then use it for concavity computation. It is not efficient and needs to be improved.

sloriot replied:
First, you don't need to copy the mesh, the `Face_filtered_graph` could be used directly in `Concavity`. Then you might even get rid of it by simply collecting the vertices per cluster if this is faster.

sloriot replied:
Actually you should precompute the convex hull to avoid computing it over and over while calling `update_edge` in your the for loops.

liubsp replied:
Good point about not copying. I need that mesh in concavity.h for normals computation. To compute the normals you need to know adjacent faces of a vertex.

update_edge is called only when any of two smaller clusters is changed. It computes all the necessary parameters like decimation cost, concavity, and later they are used for choosing next optimal edge without again calling update_edge, which implies computation of convex hull.
I think about maintaining dynamic convex hulls or implementing custom computation of convex hull using the fact that it consists of two smaller convex hulls rather than a list of random points.

Do you know any tools for profiling the code (gcc, linux)?

sloriot replied:
What I mean, if you might don't need to create the filtered_face_graph and only use the input mesh with the selection of vertices. About convex hull, you can start by considering the points on the convex hull of step n-1 + those entering the cluster (this will already filter out some useless points). `gprof` for the profiling. There should also be some nice gui on top of it.

liubsp replied:
1. Okay, I understand you.
2. I already do so :)
3. Thanks.

sloriot commented:
Please add comments in your code. I'd like to avoid reverse engineering the code.

liubsp committed:
Improve Polyhedron support, more code cleanup

liubsp committed:
Add complete Polyhedron support

liubsp committed:
Add more comments

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+
+        // main loop that stops when all edges with concavities lower that `concavity_threshold` (valid edges)
+        // are decimated or the constraint of minimum number of cluster is met
+        while (num_vertices(m_graph) > min_number_of_clusters)
```
you need to make sure that `num_vertices` is tight

sloriot replied:
min number of clusters? Aren't you looking for a maximum number?

liubsp replied:
We are looking for the minimum number of clusters that satisfy the `concavity_threshold`. On each iteration of the loop, `num_vertices` (the number of clusters) is decreased by one.

sloriot replied:
At the beginning of the algorithm each face defines a cluster but you want to merge clusters so that the number is below a given bound which is the maximum number of threshold desired.

liubsp replied:
Ok, I understand you.

liubsp replied:
But no, we cannot gurantee the threshold of maximum number of clusters is met. When there is no valid edges to decimate anymore, it stops with more clusters than that threshold. I think too leave min_number of clusters and add precondition that the number of faces must be larger or equal to the threshold.

sloriot commented on Approximate_convex_decomposition/include/CGAL/approx_decomposition.h:
```diff
@@ -0,0 +1,88 @@
+ * @return concativity value.
+ */
+template <class TriangleMesh, class FacePropertyMap,
+          class GeomTraits = CGAL::Exact_predicates_inexact_constructions_kernel
```
Nope, I think I already wrote that this should be handled with named parameters and VPM

liubsp replied:
I changed it to:
typename Kernel_traits<typename boost::property_traits<typename boost::property_map<TriangleMesh, boost::vertex_point_t>::type>::value_type>::Kernel
I don't understand why we need NamedParameters template argument

sloriot replied:
This simplifies the API for the user. Having a default value for the GeomTraits that might actually be incompatible with the graph type is not the simplest choice for the user. If GeomTraits is picked up by default based on TriangleMesh you remove the incompatibility issue.

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+        CGAL_assertion(cluster_2_props.conv_hull_pts.size() >= 3);
+        
+        // add the convex hull points of the first cluster
+        BOOST_FOREACH(Point_3 p, cluster_1_props.conv_hull_pts)
```
std::copy

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+        }
+        // add the convex hull points of the second cluster
+        BOOST_FOREACH(Point_3 p, cluster_2_props.conv_hull_pts)
+        {
```
std::copy

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+        CGAL_assertion(m_decimation_map[edge].new_cluster_props.conv_hull_pts.size() > 3);
+        
+        // assign cluster properties of the produced cluster to the first vertex of the edge
+        m_cluster_map[vert_1] = m_decimation_map[edge].new_cluster_props;
```
no need to clean `m_cluster_map` for `vert_2`?

liubsp replied:
m_cluster_map is the vertex property map of the adjacency list. I think BGL should implicitly clear its properties in `remove_vertex`, doesn't it?

sloriot replied:
OK, we're good then.

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+        remove_vertex(vert_2, m_graph);
+
+        // update decimation costs of all modified edges (only adjacent edges to the first vertex)
+        int cnt = 0;
```
cnt should also be conditional to CGAL_APPROX_DECOMPOSITION_VERBOSE

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+
+        std::set<edge_descriptor> used_edges;
+
+        BOOST_FOREACH(face_descriptor face, faces)
```
use [`border_halfedges()`](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PkgPolygonMeshProcessing.html#gaf5f5eb322811f24573aa9158e6df8070)

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/concavity.h:
```diff
@@ -0,0 +1,162 @@
+            Filtered_graph filtered_mesh(m_mesh, cluster_id, facet_ids);
+
+            TriangleMesh cluster;
+            CGAL::copy_face_graph(filtered_mesh, cluster);
```
still copying?

liubsp replied:
It doesn't compile if I try to use `filtered_mesh` instead of the copy.
56             Filtered_graph filtered_mesh(m_mesh, cluster_id, facet_ids);
 57 
 58 //            TriangleMesh cluster;
 59 //            CGAL::copy_face_graph(filtered_mesh, cluster);
 60 
 61             Concavity concavity(filtered_mesh, m_traits);
 62             return concavity.compute();
![ubuntu-2018-06-16-01-24-08](https://user-images.githubusercontent.com/12598936/41491812-01c840b4-7104-11e8-9445-9fe03ea4803e.png)

sloriot replied:
If you read the error message, you'll notice that the function you are trying to call is expecting a fixed graph type that is obviously different from the on you pass. You need to update the function signature or make it template.

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/concavity.h:
```diff
@@ -0,0 +1,162 @@
+                    if (p)
+                    {
+                        result = std::max(result, CGAL::squared_distance(origin, *p));
+                    }
```
else ?

liubsp replied:
No intersection. Actually, it should never happen because each point is projected onto the convex hull which contains them inside, so I think there is no need to add infinite concavity value.

liubsp replied:
Hmm, but it's checked in the previous if statement. This one should be removed.

sloriot replied:
The intersection of a ray and a triangle is either a segment or a point. If the normal is noisy, you might endup having a segment as intersection. In such a case I suggest to simply ignore the vertex contribution.

liubsp replied:
Ok, then I think I did fine at the beginning (checked for intersection and then checked if it's a point).

sloriot replied:
then you need to update the code

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+        {
+            decimation_props.new_cluster_props.conv_hull_pts.push_back(get(CGAL::vertex_point, conv_hull)[vert]);
+        }
+        CGAL_warning(decimation_props.new_cluster_props.conv_hull_pts.size() > 3);
```
what happen if you have some degenerate faces at initialization?

liubsp replied:
Problems might occur with convex hull computation. Can we get rid of such faces in the beginning?

sloriot replied:
if you have only 2 points the concavity measure is trivial

sloriot commented on Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h:
```diff
@@ -0,0 +1,429 @@
+        const Mesh& m_mesh;
+    };
+
+    struct Cluster_properties
```
Comment me: _struct used when evaluating the weight of collapsing an edge...._

liubsp replied:
Exactly. It contains all the necessary information about a cluster. While computing the weight (decimation cost) of an edge, information about the merged cluster is computed and is stored in the same struct inside Decimation_properties.

sloriot replied:
I mean add comment in the code to avoid reverse engineering

sloriot reviewed (commented):
Could you add me as a collaborator so that I can push in your repo?

liubsp committed:
Add fixes for remarks

liubsp committed:
Add overloaded functions for NamedParameters

liubsp committed:
Add fix for convex hull warnings

liubsp committed:
Add removal of redundant edges after decimation

liubsp committed:
Add cluster ids to properties

liubsp committed:
Add ordered list of candidate edges

liubsp committed:
Update comments

liubsp committed:
Add parallelism with TBB

liubsp committed:
Improve backward compatibility

liubsp committed:
Add plugin, fix bug with candidate edges

Plugin:
- Segmentation generation and its colorization.
- Decomposition parameters.
Bug:
- Candidates list (set) had wrong comparator, which leads to unwanted removal of candidates.
- It's strange that it worked in examples at all.

liubsp committed:
Improve the plugin's interface

sloriot committed:
make the plugin use TBB if available

sloriot committed:
reorder template parameters + use conditional value for the tags

sloriot committed:
using vpm

liubsp committed:
Add decomposition extraction

liubsp committed:
Add concavity colors to the plugin

liubsp committed:
Replace remaining tabs with spaces

sloriot commented on Approximate_convex_decomposition/include/CGAL/approx_decomposition.h:
```diff
@@ -0,0 +1,134 @@
+#ifndef CGAL_APPROXIMATE_CONVEX_DECOMPOSITION_APPROX_DECOMPOSITION_H
```
Add missing GPL license headers to all headers. See https://doc.cgal.org/latest/Manual/devman_code_format.html#secfile_header

sloriot replied:
+ you need to add the license include. I suggest that we integrate your work into the Surface mesh segmentation package because in the end, this is a segmentation method. See how a header of this package looks like to mimic it.

sloriot commented:
Before doing any further optimization here are the things to do first:
- [ ] merge this package with the surface mesh segmentation and make the functions appear in the reference manual
- [ ] update the user manual to mention also the new segmentation method
- [ ] have a testsuite that is checking the API is fully tested + some checks on precomputed example + ...
- [ ] If you look at the current function in the segmentation package, you will see that it is based on the shape diameter function (SDF). It is basically the internal diameter of a shape at each face. I'd be interested to see an example that is using the concavity of the input model instead of the SDF. See [this example](https://doc.cgal.org/latest/Surface_mesh_skeletonization/Surface_mesh_skeletonization_2segmentation_example_8cpp-example.html) where the SDF was replaced with the distance to the curve skeleton.

sloriot commented:
The doc to write documentation is here:
 https://github.com/CGAL/cgal/wiki/Documentation-Guidelines

liubsp committed:
Add GPL license

liubsp committed:
merge packages + documentation integration

liubsp committed:
update reference manual descriptions

liubsp committed:
polish reference manual

liubsp committed:
merge the folders of examples

liubsp committed:
add tests

liubsp committed:
clean up examples

liubsp committed:
remove unused meshes

liubsp committed:
update user manual, add example

liubsp committed:
improve user manual

liubsp committed:
add segmentation using concavity values example

sloriot committed:
fix bibtex ref

sloriot committed:
convex_decomposition -> approximate_convex_decomposition

sloriot committed:
fix warning

sloriot committed:
move protect macros

sloriot committed:
add Convex_hull_3 as dependencies

sloriot committed:
update doc

sloriot committed:
rename header

sloriot committed:
renaming + ref man update

sloriot committed:
do not use an unordered map if not used

liubsp committed:
update manuals

liubsp committed:
replace 4 spaces by 2

liubsp committed:
build fix

sloriot committed:
decomposition -> segmentation

sloriot committed:
add named parameter for minimum number of segments

sloriot commented on Surface_mesh_segmentation/include/CGAL/approximate_convex_decomposition.h:
```diff
@@ -0,0 +1,242 @@
+ *
+ * @param mesh triangle mesh on which approximate convex decomposition is computed
+ * @param face_ids property map with per face segment ids
+ * @param concavity_threshold maximal concavity value of the set of faces in a segment
```
Some insights must be given to the user: range of values and the effect on the result.

sloriot commented on BGL/doc/BGL/NamedParameters.txt:
```diff
@@ -114,6 +114,21 @@ as result of a partition algorithm; if it is not provided, this information is
 simply inaccessible.
 \cgalNPEnd
 
+\cgalNPBegin{minimum_number_of_segments} \anchor BGL_minimum_number_of_segments
```
I just looked at boost, and they are using max instead of maximum so we should use min here.

liubsp committed:
replace 4 spaces by 2 in the plugin

liubsp committed:
refactoring + update the plugin

liubsp committed:
rename define for verbose output

liubsp committed:
add visualization of concavity values

liubsp committed:
add shortest distance computations

liubsp committed:
fix a bug with concavity values

liubsp committed:
minimum -> min

sloriot committed:
aspell check + reworking

liubsp committed:
add convex hulls named parameter

liubsp committed:
update image in doc

liubsp committed:
add shortest distances method parameter

liubsp committed:
Merge commit 'refs/pull/1/head' of github.com:CUBERLEON/cgal into gsoc2018-approx_conv_decomp-cuberleon

liubsp committed:
update plugin and doc

liubsp committed:
add shortest concavity values image

liubsp committed:
remove unnecessary example

liubsp committed:
update convex hulls example

sloriot committed:
pass on the dc

sloriot commented on Surface_mesh_segmentation/include/CGAL/internal/Approximate_convex_segmentation/Approx_segmentation.h:
```diff
@@ -0,0 +1,582 @@
+// Author(s)     : Liubomyr Piadyk
+
+#ifndef CGAL_APPROX_SEGMENTATION_H
+#define CGAL_APPROX_SEGMENTATION_H
```
`_INTERNAL_`

liubsp committed:
Merge commit 'refs/pull/1/head' of github.com:CUBERLEON/cgal into gsoc2018-approx_conv_decomp-cuberleon

liubsp committed:
add _INTERNAL_ and a check function

liubsp committed:
add missing named params + clean up

liubsp committed:
add segments postprocessing + bug fix

- merge segments that satisfy small segment constraint
- in parallel mode there were random segmetation fault caused by racing
condition while computing normals

liubsp committed:
add benchmarks + update manuals

liubsp committed:
update user manual

sloriot commented:
I close this PR as we will open an official one.
